fix: allow exit spans to have child spans with the same type and subtype

kruskall opened the pull request:
Do not silently drop child spans of exit spans. The spec allows for exit spans to have child spans if they have the same type and subtype. One example of this is http trace requests being child spans of an exit span.

Update exit span test to make sure the feature is working as intended.

Closes https://github.com/elastic/apm-agent-go/issues/1319

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1320/detail/PR-1320/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1320/detail/PR-1320/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1320/detail/PR-1320/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1320/detail/PR-1320/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-go_1320.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-09-22T13:32:54.546Z&rangeTo=2022-09-22T13:52:54.546Z&transactionName=BUILD+apm-agent-go%2Fapm-agent-go-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=60b97035d0183b8991578df720b5a6a2&transactionId=2d7961b74a61cd46)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-09-22T13:42:54.546+0000


* Duration: 24 min 24 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2991  |
| Skipped      | 81 |
| Total        | 3072   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![9](https://img.shields.io/badge/9%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1320/detail/PR-1320/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 22 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/203/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 5 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/232/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/246/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 24 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/217/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/287/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/294/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 20 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/214/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/253/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1320/runs/1/steps/267/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `run benchmark tests` : Run the benchmark test.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

axw commented on span.go:
```diff
@@ -65,10 +65,25 @@ func (tx *Transaction) StartSpan(name, spanType string, parent *Span) *Span {
 	}
 
+	if opts.parent.IsExitSpan() {
+		arr := strings.SplitN(spanType, ".", 3)
+
+		spanType := arr[0]
+		subType := ""
+
+		if len(arr) > 1 {
+			subType = arr[1]
+		}
```
There's already code for splitting type/subtype/action. Should we extract that into a function that can be reused here?

https://github.com/elastic/apm-agent-go/blob/31ed25dde1519583f6a6160f47976be428bace87/span.go#L240-L246

kruskall replied:
This no longer applies

axw reviewed (commented):
Should we be checking the type/subtype in `Span.End`, rather than at creation time? Type/Subtype are mutable.

Also, should we check that the child span is not an exit span? The spec says that the child span must not have any destination context. That's to ensure we don't count both the exit span and its children in destination metrics.

What should happen if a child is created for a child of an exit span?

kruskall commented:
> Should we be checking the type/subtype in `Span.End`, rather than at creation time? Type/Subtype are mutable.

Good point! Changed.


> The spec says that the child span must not have any destination context. That's to ensure we don't count both the exit span and its children in destination metrics.

Updated the code to remove destination context :+1:  

> Also, should we check that the child span is not an exit span? 
> What should happen if a child is created for a child of an exit span?

The spec does not forbid that. The status of the child span has no effect on the decision of dropping the span.

axw commented on span.go:
```diff
@@ -345,6 +346,14 @@ func (s *Span) End() {
+			return
+		}
+
+		s.Context.destination = model.DestinationSpanContext{}
```
```suggestion
		s.Context.model.Destination = nil
		s.Context.model.Service = nil
```

kruskall commented on span.go:
```diff
@@ -345,6 +346,14 @@ func (s *Span) End() {
 	}
+	if s.parent.IsExitSpan() {
+		if s.Type != s.parent.Type || s.Subtype != s.parent.Subtype {
+			s.tracer = nil
```
Good point! I think we can just follow the normal control flow. I've updated the code
